snuba/query/snql/parser.py

```diff
@@ -1011,18 +1031,24 @@ def _unpack_array_conditions(
     entity_alias: Optional[str] = None,
 ) -> None:
     array_columns: Set[str] = set()
-    array_join_col = query.get_arrayjoin()
-    array_join = ""
-    if array_join_col is not None:
-        assert isinstance(array_join_col, Column)
-        array_join = f"{array_join_col.table_name + '.' if array_join_col.table_name else ''}{array_join_col.column_name}"
+    array_join_arguments = query.get_arrayjoin()
+    array_join_columns = set()
+    if array_join_arguments is not None:
+        for array_join_col in array_join_arguments:
+            assert isinstance(array_join_col, Column)
+            array_join_columns.add(
+                f"{array_join_col.table_name + '.' if array_join_col.table_name else ''}{array_join_col.column_name}"
+            )
 
     entity_alias = f"{entity_alias}." if entity_alias is not None else ""
     for column in schema:
         if isinstance(column.type, Array):
             aliased_base_name = f"{entity_alias}{column.base_name}"
             aliased_flattened = f"{entity_alias}{column.flattened}"
-            if aliased_base_name == array_join or aliased_flattened == array_join:
+            if (
+                aliased_base_name in array_join_columns
+                or aliased_flattened in array_join_columns
+            ):
```

Comment:
I am not super confident about my logic changes here so it would be good if somebody could think through potential implications

Reply:
This should work fine. Unpacking array conditions is syntactic sugar we added to make writing queries on the product side easier. If they provide a condition on an array column we will automatically convert the condition to be correct for an array column.

For example, a condition like this can be specified in the query: `tags.key = 'foo'`
In this case, the type of the column `tags.key` is an Array(String). If we passed this condition directly to Clickhouse, it would complain that you can't compare a String to an Array(String). So we automatically convert it to something of the form `arrayExists(x -> assumeNotNull(equals(x, 'foo')), tags.key)`, which is equivalent to `'foo' in tags.key`.

However, if the query is also array joining on `tags.key`, then the array column is already being unpacked, so we can pass the condition `tags.key = 'foo'` directly to Clickhouse.